# Data Cleanup of Heart Disease Dataset

---

This Jupyter Notebook was created for the **CardioVision** project by *ClinData Solutions*.

### Prerequisite
The original raw data needs to be combined into a single CSV file, including the fixed Cleveland data.

### Objectives
This Jupyter Notebook focuses on:
- Preparing raw data for the **Exploratory Data Analysis (EDA)**
- Preparing raw data for **predictive machine learning**

---

### Acknowledgements
The authors of the databases have requested that any publications resulting from the use of the data include the names of the principal investigators responsible for the data collection at each institution:

1. **Hungarian Institute of Cardiology, Budapest:** *Andras Janosi, M.D.*
2. **University Hospital, Zurich, Switzerland:** *William Steinbrunn, M.D.*
3. **University Hospital, Basel, Switzerland:** *Matthias Pfisterer, M.D.*
4. **V.A. Medical Center, Long Beach and Cleveland Clinic Foundation:** *Robert Detrano, M.D., Ph.D.*

### Original Publication
The first publication that used this dataset is:

*Detrano R, Janosi A, Steinbrunn W, Pfisterer M, Schmid JJ, Sandhu S, et al. International application of a new probability algorithm for the diagnosis of coronary artery disease. The American Journal of Cardiology. 1989; 64(5):304–10.*

---

# Table of contents

Following parts are covered in this Jupyter Notebook:
1. Importing the raw data csv
2. Selecting relevant features based on Expert opinion
3. Handling Feature Categories
4. Handling Null-Values and Duplicates

## 0. Preparations

Load all needed libraries used in this Jupyter Notebook.

In [1]:
import pandas as pd
import numpy as np

## 1. Importing the raw data csv

In [2]:
# Load the "combined_data_all.csv" and save it into a dataframe
combined_data_all_df = pd.read_csv("combined_data_all.csv")

# Show the dataframe
combined_data_all_df

Dataset   id  ccf  age  sex  painloc  painexer  relrest  pncaden  \
0        Cleveland    1    0   63    1       -9        -9       -9       -9   
1        Cleveland    2    0   67    1       -9        -9       -9       -9   
2        Cleveland    3    0   67    1       -9        -9       -9       -9   
3        Cleveland    4    0   37    1       -9        -9       -9       -9   
4        Cleveland    6    0   41    0       -9        -9       -9       -9   
..             ...  ...  ...  ...  ...      ...       ...      ...      ...   
894  Long Beach VA  200    0   54    0        1         1        1       -9   
895  Long Beach VA  201    0   62    1        0         0        0       -9   
896  Long Beach VA  202    0   55    1        1         1        1       -9   
897  Long Beach VA  116    0   58    1        1         1        1       -9   
898  Long Beach VA  160    0   62    1        1         0        0       -9   

     cp  ...  rcaprox  rcadist  lvx1  lvx2  lvx3  lvx4  lvf  cathef  junk  \
0     1  ...        1        1     1     1     1     1    1   -9.00  -9.0   
1     4  ...        1        1     1     1     1     1    1   -9.00  -9.0   
2     4  ...        2        2     1     1     1     7    3   -9.00  -9.0   
3     3  ...        1        1     1     1     1     1    1   -9.00  -9.0   
4     2  ...        1        1     1     1     1     1    1   -9.00  -9.0   
..   ..  ...      ...      ...   ...   ...   ...   ...  ...     ...   ...   
894   4  ...        1        1     1     1     1     1    1    0.76   5.6   
895   1  ...        1        1     1     1     1     1    2    0.62   3.5   
896   4  ...        2        1     1     1     1     1    1    0.69   5.6   
897   4  ...        1        1     1     1     1     1    1    0.81   6.0   
898   2  ...        2        1     1     1     1     1    1   -9.00  -9.0   

     name  
0    name  
1    name  
2    name  
3    name  
4    name  
..    ...  
894  name  
895  name  
896  name  
897  name  
898  name  

[899 rows x 77 columns]

## 2. Selection of Chosen Variables

The original dataset contains 76 features, but only 14 have been widely used in publications so far. Some features describe others; for instance, the boolean "hypertension" feature indicates whether the resting blood pressure value is above or below a certain threshold.

All 14 commonly used features were also identified by our domain expert as they are known risk factors for heart disease. In addition, 4 more features were selected for inclusion in the initial Exploratory Data Analysis (EDA).

### Selected Features Table

| **No.** | **Feature Name** | **Description**                                            | **Feature Type**         |
|---------|------------------|----------------------------------------------------------|--------------------------|
| 1       | age              | Age of the patient                                        | Numerical                |
| 2       | sex              | Sex of the patient (1 = male; 0 = female)                 | Categorical (Binary)     |
| 3       | cp               | Chest pain type (1 = typical angina; 2 = atypical angina; 3 = non-anginal pain; 4 = asymptomatic) | Categorical              |
| 4       | chol             | Serum cholesterol in mg/dl                                | Numerical                |
| 5       | fbs              | Fasting blood sugar > 120 mg/dl (1 = true; 0 = false)     | Categorical (Binary)     |
| 6       | restecg          | Resting electrocardiographic results                      | Categorical              |
| 7       | thalach          | Maximum heart rate achieved                               | Numerical                |
| 8       | exang            | Exercise-induced angina (1 = yes; 0 = no)                 | Categorical (Binary)     |
| 9       | oldpeak          | ST depression induced by exercise relative to rest       | Numerical                |
| 10      | slope            | The slope of the peak exercise ST segment (1 = upsloping; 2 = flat; 3 = downsloping) | Categorical              |
| 11      | ca               | Number of major vessels (0-3) colored by fluoroscopy     | Numerical (Discrete)     |
| 12      | painloc          | Chest pain location (1 = substernal; 0 = otherwise)       | Categorical (Binary)     |
| 13      | htn              | History of hypertension (1 = yes; 0 = no)                 | Categorical (Binary)     |
| 14      | cigs             | Number of cigarettes smoked per day                      | Numerical                |
| 15      | years            | Number of years as a smoker                              | Numerical                |
| 16      | famhist          | Family history of coronary artery disease (1 = yes; 0 = no) | Categorical (Binary)  |
| 17      | thalrest         | Resting heart rate                                       | Numerical                |
| 18      | num              | Diagnosis of heart disease (0-4, representing increasing severity) | Categorical (Ordinal)   |

---





In [3]:
# Create a new df, that keeps only the selected columns.
selected_df = combined_data_all_df[["age", "sex", "cp", "chol", "fbs", "restecg", "thalach", "exang", "oldpeak", "slope", "ca", "painloc", "htn", 
 "cigs", "years", "famhist", "thalrest", "num"]]

# Display the new df.
selected_df

age  sex  cp  chol  fbs  restecg  thalach  exang  oldpeak  slope  ca  \
0     63    1   1   233    1        2      150      0      2.3      3   0   
1     67    1   4   286    0        2      108      1      1.5      2   3   
2     67    1   4   229    0        2      129      1      2.6      2   2   
3     37    1   3   250    0        0      187      0      3.5      3   0   
4     41    0   2   204    0        2      172      0      1.4      1   0   
..   ...  ...  ..   ...  ...      ...      ...    ...      ...    ...  ..   
894   54    0   4   333    1        1      154      0      0.0     -9  -9   
895   62    1   1   139    0        1       -9     -9     -9.0     -9  -9   
896   55    1   4   223    1        1      100      0      0.0     -9  -9   
897   58    1   4   385    1        2       -9     -9     -9.0     -9  -9   
898   62    1   2   254    0        2       93      1      0.0     -9  -9   

     painloc  htn  cigs  years  famhist  thalrest  num  
0         -9    1    50     20        1        60    0  
1         -9    1    40     40        1        64    2  
2         -9    1    20     35        1        78    1  
3         -9    0     0      0        1        84    0  
4         -9    1     0      0        1        71    0  
..       ...  ...   ...    ...      ...       ...  ...  
894        1    0     0      0        1        83    1  
895        0    0    15     30        0        -9    0  
896        1    1    20     40        0        74    2  
897        1    0    10     20        1        -9    0  
898        1    1     0      0        0        67    1  

[899 rows x 18 columns]

## 3. Handling Feature Categories

Check all datatypes of the features and change them accordingly to the table above.

In [4]:
# Check the datatypes of the selected data
selected_df.dtypes

age           int64
sex           int64
cp            int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
painloc       int64
htn           int64
cigs          int64
years         int64
famhist       int64
thalrest      int64
num           int64
dtype: object

### Change data types and NaN
The datatype of some features is int64 instead of categorical, as mentioned in the table above. 
They should be changed to optimize memory usage and for the machine learning algorithms.

!!! Attention !!!
The dataset uses -9 for NaN values. These should be replaced by NaN again, as numpy and pandas are working with NaN natively.
This needs to be done before changing the datatype to categorical.

In [5]:
# Replace -9 with NaN
selected_df = selected_df.replace(-9, np.nan)

### Change the numeric value to actual category
The numeric value should be changed to the actual category to make the analytics better understandable.

In [6]:
# Covnert sex to descriptive labels
selected_df['sex'] = selected_df['sex'].replace({1: 'Male', 0: 'Female'})

# Convert 'cp' (chest pain) to descriptive labels
selected_df['cp'] = selected_df['cp'].replace({
    1: 'Typical Angina',
    2: 'Atypical Angina',
    3: 'Non-Anginal Pain',
    4: 'Asymptomatic'
})

# Convert 'fbs' (fasting blood sugar > 120 mg/dl) to True/False
selected_df['fbs'] = selected_df['fbs'].replace({1: True, 0: False})

# Convert 'restecg' to descriptive labels
selected_df['restecg'] = selected_df['restecg'].replace({
    0: 'Normal',
    1: 'ST-T Wave Abnormality',
    2: 'Left Ventricular Hypertrophy'
})

# Convert 'exang' (exercise induced angina) to True/False
selected_df['exang'] = selected_df['exang'].replace({1: True, 0: False})

# Convert 'slope' to descriptive labels
selected_df['slope'] = selected_df['slope'].replace({
    1: 'Upsloping',
    2: 'Flat',
    3: 'Downsloping'
})

# Convert 'painloc' (chest pain location) to descriptive labels
selected_df['painloc'] = selected_df['painloc'].replace({1: 'Substernal', 0: 'Otherwise'})

# Convert 'famhist' (family history of coronary artery disease) to True/False
selected_df['famhist'] = selected_df['famhist'].replace({1: True, 0: False})

selected_df

age     sex                cp   chol    fbs  \
0     63    Male    Typical Angina  233.0   True   
1     67    Male      Asymptomatic  286.0  False   
2     67    Male      Asymptomatic  229.0  False   
3     37    Male  Non-Anginal Pain  250.0  False   
4     41  Female   Atypical Angina  204.0  False   
..   ...     ...               ...    ...    ...   
894   54  Female      Asymptomatic  333.0   True   
895   62    Male    Typical Angina  139.0  False   
896   55    Male      Asymptomatic  223.0   True   
897   58    Male      Asymptomatic  385.0   True   
898   62    Male   Atypical Angina  254.0  False   

                          restecg  thalach  exang  oldpeak        slope   ca  \
0    Left Ventricular Hypertrophy    150.0  False      2.3  Downsloping  0.0   
1    Left Ventricular Hypertrophy    108.0   True      1.5         Flat  3.0   
2    Left Ventricular Hypertrophy    129.0   True      2.6         Flat  2.0   
3                          Normal    187.0  False      3.5  Downsloping  0.0   
4    Left Ventricular Hypertrophy    172.0  False      1.4    Upsloping  0.0   
..                            ...      ...    ...      ...          ...  ...   
894         ST-T Wave Abnormality    154.0  False      0.0          NaN  NaN   
895         ST-T Wave Abnormality      NaN    NaN      NaN          NaN  NaN   
896         ST-T Wave Abnormality    100.0  False      0.0          NaN  NaN   
897  Left Ventricular Hypertrophy      NaN    NaN      NaN          NaN  NaN   
898  Left Ventricular Hypertrophy     93.0   True      0.0          NaN  NaN   

        painloc  htn  cigs  years famhist  thalrest  num  
0           NaN  1.0  50.0   20.0    True      60.0    0  
1           NaN  1.0  40.0   40.0    True      64.0    2  
2           NaN  1.0  20.0   35.0    True      78.0    1  
3           NaN  0.0   0.0    0.0    True      84.0    0  
4           NaN  1.0   0.0    0.0    True      71.0    0  
..          ...  ...   ...    ...     ...       ...  ...  
894  Substernal  0.0   0.0    0.0    True      83.0    1  
895   Otherwise  0.0  15.0   30.0   False       NaN    0  
896  Substernal  1.0  20.0   40.0   False      74.0    2  
897  Substernal  0.0  10.0   20.0    True       NaN    0  
898  Substernal  1.0   0.0    0.0   False      67.0    1  

[899 rows x 18 columns]

In [8]:
# Convert columns to categorical data types, handling NaN values properly
columns_to_convert = ['sex', 'cp', 'fbs', 'restecg', 'exang', 'slope', 'ca', 'painloc', 'htn', 'famhist']

for col in columns_to_convert:
    selected_df[col] = selected_df[col].astype("category")

# Display the data types to confirm the changes
print(selected_df.dtypes)

age            int64
sex         category
cp          category
chol         float64
fbs         category
restecg     category
thalach      float64
exang       category
oldpeak      float64
slope       category
ca          category
painloc     category
htn         category
cigs         float64
years        float64
famhist     category
thalrest     float64
num            int64
dtype: object


### Rename columns for better understanding
| **Current Column**   | **Description**                                             | **New Column Name**            |
|----------------------|-------------------------------------------------------------|--------------------------------|
| `age`                | Age in years                                                | `age`                          |
| `sex`                | Sex (1 = male; 0 = female)                                  | `gender`                       |
| `cp`                 | Chest pain type                                             | `chest_pain_type`              |
| `chol`               | Serum cholesterol in mg/dl                                  | `serum_cholesterol`            |
| `fbs`                | Fasting blood sugar > 120 mg/dl (1 = true; 0 = false)       | `high_fasting_blood_sugar`     |
| `restecg`            | Resting electrocardiographic results                        | `resting_ecg_results`          |
| `thalach`            | Maximum heart rate achieved                                 | `max_heart_rate`               |
| `exang`              | Exercise induced angina (1 = yes; 0 = no)                   | `exercise_induced_angina`      |
| `oldpeak`            | ST depression induced by exercise relative to rest          | `st_depression`                |
| `slope`              | The slope of the peak exercise ST segment                   | `st_slope`                     |
| `ca`                 | Number of major vessels (0-3) colored by fluoroscopy        | `num_major_vessels`            |
| `painloc`            | Chest pain location                                         | `chest_pain_location`          |
| `htn`                | History of hypertension (1 = yes; 0 = no)                   | `has_hypertension`             |
| `cigs`               | Cigarettes smoked per day                                   | `cigarettes_per_day`           |
| `years`              | Number of years as a smoker                                 | `years_smoking`                |
| `famhist`            | Family history of coronary artery disease (1 = yes; 0 = no) | `family_history_cad`           |
| `thalrest`           | Resting heart rate                                          | `resting_heart_rate`           |
| `num`                | Diagnosis of heart disease                                  | `heart_disease_diagnosis`      |


In [15]:
# Rename columns based on the provided documentation
selected_df = selected_df.rename(columns={
    'age': 'age',
    'sex': 'gender',
    'cp': 'chest_pain_type',
    'chol': 'serum_cholesterol',
    'fbs': 'high_fasting_blood_sugar',
    'restecg': 'resting_ecg_results',
    'thalach': 'max_heart_rate',
    'exang': 'exercise_induced_angina',
    'oldpeak': 'st_depression',
    'slope': 'st_slope',
    'ca': 'num_major_vessels',
    'painloc': 'chest_pain_location',
    'htn': 'has_hypertension',
    'cigs': 'cigarettes_per_day',
    'years': 'years_smoking',
    'famhist': 'family_history_cad',
    'thalrest': 'resting_heart_rate',
    'num': 'heart_disease_diagnosis'
})

selected_df

age  gender   chest_pain_type  serum_cholesterol  \
0     63    Male    Typical Angina              233.0   
1     67    Male      Asymptomatic              286.0   
2     67    Male      Asymptomatic              229.0   
3     37    Male  Non-Anginal Pain              250.0   
4     41  Female   Atypical Angina              204.0   
..   ...     ...               ...                ...   
894   54  Female      Asymptomatic              333.0   
895   62    Male    Typical Angina              139.0   
896   55    Male      Asymptomatic              223.0   
897   58    Male      Asymptomatic              385.0   
898   62    Male   Atypical Angina              254.0   

    high_fasting_blood_sugar           resting_ecg_results  max_heart_rate  \
0                       True  Left Ventricular Hypertrophy           150.0   
1                      False  Left Ventricular Hypertrophy           108.0   
2                      False  Left Ventricular Hypertrophy           129.0   
3                      False                        Normal           187.0   
4                      False  Left Ventricular Hypertrophy           172.0   
..                       ...                           ...             ...   
894                     True         ST-T Wave Abnormality           154.0   
895                    False         ST-T Wave Abnormality             NaN   
896                     True         ST-T Wave Abnormality           100.0   
897                     True  Left Ventricular Hypertrophy             NaN   
898                    False  Left Ventricular Hypertrophy            93.0   

    exercise_induced_angina  st_depression     st_slope num_major_vessels  \
0                     False            2.3  Downsloping               0.0   
1                      True            1.5         Flat               3.0   
2                      True            2.6         Flat               2.0   
3                     False            3.5  Downsloping               0.0   
4                     False            1.4    Upsloping               0.0   
..                      ...            ...          ...               ...   
894                   False            0.0          NaN               NaN   
895                     NaN            NaN          NaN               NaN   
896                   False            0.0          NaN               NaN   
897                     NaN            NaN          NaN               NaN   
898                    True            0.0          NaN               NaN   

    chest_pain_location has_hypertension  cigarettes_per_day  years_smoking  \
0                   NaN              1.0                50.0           20.0   
1                   NaN              1.0                40.0           40.0   
2                   NaN              1.0                20.0           35.0   
3                   NaN              0.0                 0.0            0.0   
4                   NaN              1.0                 0.0            0.0   
..                  ...              ...                 ...            ...   
894          Substernal              0.0                 0.0            0.0   
895           Otherwise              0.0                15.0           30.0   
896          Substernal              1.0                20.0           40.0   
897          Substernal              0.0                10.0           20.0   
898          Substernal              1.0                 0.0            0.0   

    family_history_cad  resting_heart_rate  heart_disease_diagnosis  
0                 True                60.0                        0  
1                 True                64.0                        2  
2                 True                78.0                        1  
3                 True                84.0                        0  
4                 True                71.0                        0  
..                 ...                 ...                      ...  
894               T

### Modification of Target Variable
The current target feature, num (heart disease diagnosis), ranges from 0 to 4, where 0 indicates no heart disease, and 1-4 represent increasing levels of heart disease severity.

To determine whether a binary classification (0 = no heart disease, 1 = heart disease) or a multiclass classification approach is more suitable, the distribution of these classes will be analyzed.

In [17]:
# Checking the distribution of values in the target column
distribution = selected_df['heart_disease_diagnosis'].value_counts().sort_index()

# Calculate the percentage distribution
percentage_distribution = (distribution / len(selected_df)) * 100

percentage_distribution

heart_disease_diagnosis
0    44.938821
1    21.245829
2    14.460512
3    14.682981
4     4.671858
Name: count, dtype: float64

Due to the observed class imbalance, particularly for num = 4, we will begin with a binary classification approach. All heart disease severity levels (1-4) will be consolidated into a single category labeled as heart disease = True, while 0 will represent heart disease = False.

In [20]:
# Convert heart disease severity into a binary classification
selected_df['heart_disease_diagnosis'] = selected_df['heart_disease_diagnosis'].apply(lambda x: True if x > 0 else False)

# Verify the transformation
selected_df['heart_disease_diagnosis'].value_counts()


heart_disease_diagnosis
True     495
False    404
Name: count, dtype: int64

The target values are now balanced, with 495 patients classified as having heart disease and 404 patients as not having heart disease. For the time being, this balanced dataset will be used for binary classification.

## 4. Handle missing and duplicated values

The Null/ NaN values were originally replaced by -9. Before converting the to categorical features, the -9 was again replaced by NaN so that the built-in functions are working again.

In [22]:
# Calculate the total number of missing values per feature
missing_values_total = selected_df.isnull().sum()

# Calculate the percentage of missing values per feature
missing_values_percentage = (missing_values_total / len(selected_df)) * 100

# Create a DataFrame to display both total and percentage of missing values per feature
missing_values_table = pd.DataFrame({
    'Total Missing': missing_values_total,
    'Percentage Missing (%)': missing_values_percentage
})

# Display the table
print(missing_values_table)

                          Total Missing  Percentage Missing (%)
age                                   0                0.000000
gender                                0                0.000000
chest_pain_type                       0                0.000000
serum_cholesterol                    30                3.337041
high_fasting_blood_sugar             90               10.011123
resting_ecg_results                   2                0.222469
max_heart_rate                       55                6.117909
exercise_induced_angina              55                6.117909
st_depression                        62                6.896552
st_slope                            308               34.260289
num_major_vessels                   608               67.630701
chest_pain_location                 282               31.368187
has_hypertension                     34                3.781980
cigarettes_per_day                  420               46.718576
years_smoking                       432 

### Handle Null Values

We need to find a way to handle the features with high amounts of null values such as ca, cigs and so so on...

### Handle Duplicates
Check for duplicates in the data and handle them accordingly.

In [10]:
# Check the dataset for duplicates, if false continue
selected_df.duplicated().any()

np.True_

In [11]:
# Identify the duplicate rows
duplicate_row = selected_df[selected_df.duplicated(keep=False)]

# Display the rows that are identical
print(duplicate_row)

     age   sex                cp   chol    fbs                restecg  \
838   58  Male  Non-Anginal Pain  219.0  False  ST-T Wave Abnormality   
886   58  Male  Non-Anginal Pain  219.0  False  ST-T Wave Abnormality   

     thalach exang  oldpeak slope  ca     painloc  htn  cigs  years famhist  \
838    118.0  True      0.0   NaN NaN  Substernal  1.0  20.0   35.0    True   
886    118.0  True      0.0   NaN NaN  Substernal  1.0  20.0   35.0    True   

     thalrest  num  
838      70.0    2  
886      70.0    2  


Based on the selected features, these entries appear identical. However, when considering all features, there are differences that suggest they represent two distinct individuals. As a result, both entries will be retained in the dataset.

In [12]:
# Perform full descriptive analysis of the dataset to check if all works
selected_df.describe(include='all')

age   sex            cp        chol    fbs restecg     thalach  \
count   899.000000   899           899  869.000000    809     897  844.000000   
unique         NaN     2             4         NaN      2       3         NaN   
top            NaN  Male  Asymptomatic         NaN  False  Normal         NaN   
freq           NaN   711           485         NaN    674     538         NaN   
mean     53.480534   NaN           NaN  198.759494    NaN     NaN  137.298578   
std       9.435894   NaN           NaN  111.834415    NaN     NaN   25.965959   
min      28.000000   NaN           NaN    0.000000    NaN     NaN   60.000000   
25%      47.000000   NaN           NaN  175.000000    NaN     NaN  120.000000   
50%      54.000000   NaN           NaN  224.000000    NaN     NaN  140.000000   
75%      60.000000   NaN           NaN  269.000000    NaN     NaN  157.000000   
max      77.000000   NaN           NaN  603.000000    NaN     NaN  202.000000   

        exang     oldpeak slope     ca     painloc    htn        cigs  \
count     844  837.000000   591  291.0         617  865.0  479.000000   
unique      2         NaN     4    5.0           2    2.0         NaN   
top     False         NaN  Flat    0.0  Substernal    0.0         NaN   
freq      514         NaN   334  171.0         568  453.0         NaN   
mean      NaN    0.870490   NaN    NaN         NaN    NaN   19.118998   
std       NaN    1.080548   NaN    NaN         NaN    NaN   18.296273   
min       NaN   -2.600000   NaN    NaN         NaN    NaN    0.000000   
25%       NaN    0.000000   NaN    NaN         NaN    NaN    0.000000   
50%       NaN    0.500000   NaN    NaN         NaN    NaN   20.000000   
75%       NaN    1.500000   NaN    NaN         NaN    NaN   30.000000   
max       NaN    6.200000   NaN    NaN         NaN    NaN   99.000000   

             years famhist    thalrest         num  
count   467.000000     477  843.000000  899.000000  
unique         NaN       2         NaN         NaN  
top            NaN    True         NaN         NaN  
freq           NaN     269         NaN         NaN  
mean     18.796574     NaN   75.487544    1.129032  
std      16.359145     NaN   14.727961    1.259720  
min       0.000000     NaN   37.000000    0.000000  
25%       0.000000     NaN   65.000000    0.000000  
50%      20.000000     NaN   74.000000    1.000000  
75%      30.000000     NaN   84.000000    2.000000  
max      60.000000     NaN  139.000000    4.000000

## 5. Save the final df to a csv file that is ready to be used for the EDA and machine learning

In [13]:
selected_df.to_csv("dataset_with_selected_features.csv", index=False)